File: CSharp/Math/Tuple.cs
using System;

namespace CSharp
{
    public class Tuple
    {
        public Tuple(double x, double y, double z, double w)
        {
            this.x = x;
            this.y = y;
            this.z = z;
            this.w = w;
        }

        public double x { get; protected set; }
        public double y { get; protected set; }
        public double z { get; protected set; }
        public double w { get; protected set; }

        public bool IsPoint => w.IsEqual(1.0);
        public bool IsVector => w.IsEqual(0.0);

        public void SetX(double X) => x = X;

        public void SetY(double Y) => y = Y;

        public void SetZ(double Z) => z = Z;

        public void SetW(double W) => w = W;

        public static Tuple operator+(Tuple lhs, Tuple rhs)
        {
            return new Tuple(lhs.x + rhs.x,
                             lhs.y + rhs.y,
                             lhs.z + rhs.z,
                             lhs.w + rhs.w);
        }

        public static Tuple operator-(Tuple lhs, Tuple rhs)
        {
            return new Tuple(lhs.x - rhs.x,
                             lhs.y - rhs.y,
                             lhs.z - rhs.z,
                             lhs.w - rhs.w);
        }

        public static Tuple operator-(Tuple tuple)
        {
            return new Tuple(-tuple.x,
                             -tuple.y,
                             -tuple.z,
                             -tuple.w);
        }

        public static Tuple operator *(Tuple lhs, double rhs)
        {
            return new Tuple(lhs.x * rhs,
                             lhs.y * rhs,
                             lhs.z * rhs,
                             lhs.w * rhs);
        }

        public static Tuple operator/(Tuple lhs, double rhs)
        {
            return new Tuple(lhs.x / rhs,
                             lhs.y / rhs,
                             lhs.z / rhs,
                             lhs.w / rhs);
        }

        public override bool Equals(object obj)
        {
            var other = obj as Tuple;
            return ((other != null) &&
                    (x.IsEqual(other.x)) &&
                    (y.IsEqual(other.y)) &&
                    (z.IsEqual(other.z)) &&
                    (w.IsEqual(other.w)));
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(x, y, z, w);
        }

        public override string ToString()
        {
            return $"({x},{y},{z},{w})";
        }
    }
}

File: Java/src/main/java/Math/Tuple.java
package RTC.Math;

import java.util.Objects;

public class Tuple {
    public static int TupleId = 0;

    private int id;
    private double x, y, z, w;

    public Tuple(double X, double Y, double Z, double W) {
        id = TupleId++;

        x = X;
        y = Y;
        z = Z;
        w = W;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double getZ() {
        return z;
    }

    public double getW() {
        return w;
    }
    
    public void setW(double W) {
        w = W;
    }
    
    public boolean isPoint() {
        return Helper.IsEqual(w, 1.0);
    }

    public boolean isVector() {
        return Helper.IsEqual(w, 0.0);
    }

    public Tuple add(Tuple t) {
        return new Tuple(x + t.getX(),
                         y + t.getY(),
                         z + t.getZ(),
                         w + t.getW());
    }

    public Tuple sub(Tuple t) {
        return new Tuple(x - t.getX(),
                         y - t.getY(),
                         z - t.getZ(),
                         w - t.getW());
    }

    public Tuple neg() {
        return new Tuple(-x, -y, -z, -w);
    }

    public Tuple mul(double a) {
        return new Tuple(x * a, y * a, z * a, w * a);
    }

    public Tuple div(double a) {
        return new Tuple(x / a, y / a, z / a, w / a);
    }

    public double mag() {
        return Math.sqrt(x * x + y * y + z * z + w * w);
    }

    public Tuple norm() {
        double mag = mag();
        return new Tuple(x / mag, y / mag, z / mag, w / mag);
    }

    public double dot(Tuple t) {
        return x * t.getX() +
               y * t.getY() +
               z * t.getZ() +
               w * t.getW();
    }

    public Tuple cross(Tuple t)
    {
        return new Tuple(y * t.getZ() - z * t.getY(),
                         z * t.getX() - x * t.getZ(),
                         x * t.getY() - y * t.getX(),
                         0);
    }

    public Tuple reflect(Tuple normal) {
        double d = 2 * dot(normal);
        Tuple n = normal.mul(d);
        return sub(n);
    }

    @Override
    public boolean equals(Object o)
    {
        if (o == this) {
            return true;
        }

        if ((o == null) || !(o instanceof Tuple)) {
            return false;
        }

        Tuple t = (Tuple) o;
        return Helper.IsEqual(x, t.getX())
        && Helper.IsEqual(y, t.getY())
        && Helper.IsEqual(z, t.getZ())
        && Helper.IsEqual(w, t.getW());
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, x, y, z, w);
    }

    @Override
    public String toString() {
        return String.format("(%f,%f,%f,%f) Id: %d", x, y, z, w, id);
    }
}
